components/brave_ads/core/internal/common/url/url_util_internal.cc

```diff
@@ -0,0 +1,120 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/core/internal/common/url/url_util_internal.h"
+
+#include <string>
+
+#include "base/containers/contains.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "net/base/url_util.h"
+#include "url/gurl.h"
+
+namespace brave_ads::internal {
+
+namespace {
+
+constexpr char kBraveScheme[] = "brave";
+constexpr char kChromeScheme[] = "chrome";
+
+constexpr char kRewardsHostName[] = "rewards";
+constexpr char kSyncHostName[] = "sync";
+constexpr char kWalletHostName[] = "wallet";
+
+constexpr char kSettingsHostName[] = "settings";
+
+constexpr char kSearchEnginesPath[] = "/searchEngines";
+constexpr char kSearchPath[] = "/search";
+constexpr char kSearchQuery[] = "search";
+
+}  // namespace
+
+GURL ReplaceBraveSchemeWithChromeScheme(const GURL& url) {
+  CHECK(url.is_valid());
+
+  if (!url.SchemeIs(kBraveScheme)) {
+    return url;
+  }
+
+  GURL::Replacements replacements;
+  replacements.SetSchemeStr(kChromeScheme);
+  return url.ReplaceComponents(replacements);
+}
+
+bool HasQueryNameAndNonEmptyValue(const GURL& url,
+                                  std::string_view query_name) {
+  for (net::QueryIterator iter(url); !iter.IsAtEnd(); iter.Advance()) {
+    if (iter.GetKey() != query_name || iter.GetValue().empty()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool DoesETLDPlusOneContainWildcard(const GURL& url) {
+  constexpr char kEscapedAsteriskWildcard[] = "%2A";
+
+  const std::string domain_and_registry =
+      net::registry_controlled_domains::GetDomainAndRegistry(
+          url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  // We do not need to check for question mark wildcard because GURL strings are
+  // invalid if the domain and registry contain a question mark.
+  return base::Contains(
+      net::registry_controlled_domains::GetDomainAndRegistry(
+          url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES),
+      kEscapedAsteriskWildcard);
+}
+
+bool IsHostIcannRegistered(std::string_view host,
+                           const bool include_private_registries) {
+  return net::registry_controlled_domains::HostHasRegistryControlledDomain(
+      host,
+      /*ICANN*/ net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES,
+      include_private_registries
+          ? net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES
+          : net::registry_controlled_domains::EXCLUDE_PRIVATE_REGISTRIES);
+}
+
+bool DoesSupportInternalUrlWithBraveScheme(const GURL& url) {
```

Comment:
This function was decoupled from `url_util.[h|cc]` and had a privacy review in the past.